File: frontend/rpc.ts
import { Millennium } from "@steambrew/client";
import { logger } from "./index";

/**
 * What a backend method answers instead of a result when the work it needs is
 * too slow to hold the Lua state for. RPC handlers run on Millennium's main
 * thread and cannot yield, so the only way not to block every other call is to
 * start the work, say so, and be asked again - which is what {@link call} does,
 * invisibly to everything below.
 */
interface Pending {
  pending: true;
  /** How long the backend wants before the next attempt, in milliseconds. */
  retryIn?: number;
}

const isPending = (value: unknown): value is Pending =>
  typeof value === "object" && value !== null && (value as Pending).pending === true;

/** Stop retrying eventually, so a backend stuck saying "pending" cannot hang a
 * caller for good. */
const RETRY_LIMIT_MS = 120_000;

const sleep = (ms: number) => new Promise((resolve) => setTimeout(resolve, ms));

// The backend speaks snake_case, because that's what legendary and Lua both
// use. Converting every answer here once keeps the rest of the frontend in
// normal TS casing without a hand-written mapping per document - which was
// where a renamed backend field used to turn silently into `undefined`.
//
// Only answers are converted. Payloads still go out spelled the way the Lua
// side reads them.

function camelize(value: unknown): unknown {
  if (Array.isArray(value)) return value.map(camelize);
  if (typeof value !== "object" || value === null) return value;

  const result: Record<string, unknown> = {};
  for (const [key, nested] of Object.entries(value)) {
    result[key.replace(/_(.)/g, (_, character: string) => character.toUpperCase())] =
      camelize(nested);
  }

  return result;
}

async function call<R>(route: `RPC.${string}`, payload: object = {}): Promise<R> {
  const name = route.slice(4);
  logger.debug(`-> ${name}`, payload);

  const body = JSON.stringify(payload);
  const deadline = Date.now() + RETRY_LIMIT_MS;

  for (;;) {
    const raw = await Millennium.callServerMethod(name, { payload: body });
    const parsed = camelize(JSON.parse(raw)) as R | Pending;

    if (!isPending(parsed)) {
      logger.debug(`<- ${name}`, parsed);
      return parsed;
    }

    if (Date.now() > deadline) {
      logger.warn(`${name} never stopped saying pending`);
      return parsed as R;
    }

    await sleep(parsed.retryIn ?? 250);
  }
}

/**
 * Lua has one table type, so an empty list arrives as `{}` - an object with no
 * `map` on it, and a TypeError on every fresh install.
 */
function asArray<T>(value: T[] | undefined): T[] {
  return Array.isArray(value) ? value : [];
}

/** Whether legendary is usable, and whether an Epic account is signed in. */
export interface EpicStatus {
  available: boolean;
  authenticated: boolean;
  account?: string;
  loginUrl: string;
  error?: string;
}

/** One game the signed in Epic account owns, installed or not. */
export interface EpicGame {
  /** Epic's internal id, and the stable key everything here is stored under. */
  appName: string;
  title: string;
  installed: boolean;
  installPath?: string;
  installSize?: number;
  version?: string;
  needsUpdate: boolean;
  /** Directory Epic expects the game in, which is what legendary names it. */
  folderName?: string;
  artPortrait?: string;
  artHero?: string;
}

/** How much room a game takes, in bytes. */
export interface GameSize {
  /** On disk once installed - what Steam labels "Space Required". */
  disk: number;
  /** Actually transferred, which is smaller: Epic ships its content compressed. */
  download: number;
}

/** One Epic achievement, unlocked or not. */
export interface Achievement {
  /** Epic's internal name, unique within the game. */
  id: string;
  name: string;
  description: string;
  unlocked: boolean;
  /** 0 to 1. Only the handful of achievements Epic tracks progress for move. */
  progress: number;
  /** Unix seconds, absent while locked. */
  unlockedAt?: number;
  icon?: string;
  /** Percent of players holding it. */
  rarity?: number;
  /** Undiscovered, so Steam draws it blurred. */
  hidden: boolean;
}

/** Every achievement one game has, with the account's progress folded in. */
export interface GameAchievements {
  total: number;
  unlocked: number;
  /** Unix seconds of the read from Epic, which is what makes this stale. */
  fetchedAt: number;
  achievements: Achievement[];
}

/** What Steam needs to create a shortcut that launches a game through legendary. */
export interface LaunchCommand {
  exe: string;
  arguments: string;
  startDir: string;
}

/** How far along a running install is. Absent until legendary's first progress line. */
export interface JobProgress {
  percent: number;
  downloaded: number;
  total: number;
  /** "00:10:00", straight from legendary. */
  eta?: string;
  elapsed?: string;
  /** MiB/s, raw off the wire. */
  speed?: number;
}

/** One detached legendary install or uninstall. */
export interface Job {
  appName: string;
  kind: "install" | "uninstall";
  /**
   * `queued` is a job with no process behind it yet: legendary runs one at a
   * time, so the rest wait their turn. `paused` is a killed runner - legendary
   * has no pause, so resuming re-runs install.
   */
  state: "queued" | "running" | "paused" | "done" | "failed";
  /** When the job was asked for, which is also its place in the queue. */
  startedAt: number;
  /** When its runner was launched, absent while it is still queued. */
  spawnedAt?: number;
  exitCode?: number;
  progress?: JobProgress;
  error?: string;
}

export interface LibraryResult {
  ok: boolean;
  error?: string;
  games: EpicGame[];
  /** Unix seconds of the last read from Epic, 0 if there has never been one. */
  refreshedAt: number;
}

interface RawLibrary {
  ok: boolean;
  error?: string;
  games?: EpicGame[];
  refreshedAt?: number;
}

/** One game's achievement counts, without the achievements themselves. */
export interface AchievementSummary {
  appName: string;
  total: number;
  unlocked: number;
  fetchedAt: number;
}

type RawAchievement = Omit<Achievement, "unlockedAt"> & { unlockedAt?: string };
type RawAchievements = Omit<GameAchievements, "achievements"> & {
  achievements?: RawAchievement[];
};

/**
 * Epic times its unlocks as "2026-06-12 21:21:24.391000+00:00", which is ISO
 * 8601 with a space where the T belongs - so Date can't read it as it stands,
 * and Steam wants Unix seconds anyway.
 */
function toAchievement(raw: RawAchievement): Achievement {
  const parsed = raw.unlockedAt ? Date.parse(raw.unlockedAt.replace(" ", "T")) : Number.NaN;

  return {
    ...raw,
    unlockedAt: Number.isNaN(parsed) ? undefined : Math.floor(parsed / 1000),
  };
}

function toLibrary(raw: RawLibrary): LibraryResult {
  return { ...raw, games: asArray(raw.games), refreshedAt: raw.refreshedAt ?? 0 };
}

export class RPC {
  /**
   * Cached in the backend, because answering costs a subprocess launch; pass
   * `refresh` after anything that could have changed the answer.
   */
  async GetStatus(refresh = false) {
    return call<EpicStatus>("RPC.GetStatus", { refresh });
  }

  /**
   * Finish signing in, with the `authorizationCode` from Epic's redirect page.
   * The browser half happens in the panel - this is only the exchange.
   */
  async SignIn(code: string) {
    return call<{ ok: boolean; error?: string; status?: EpicStatus }>("RPC.SignIn", { code });
  }

  async SignOut() {
    return call<{ ok: boolean; error?: string; status?: EpicStatus }>("RPC.SignOut");
  }

  /**
   * Answered from the backend's cache unless `refresh` is set; `force` also
   * bypasses legendary's catalog cache, which is what picks up a newly bought
   * game.
   */
  async GetLibrary(refresh = false, force = false): Promise<LibraryResult> {
    return toLibrary(await call<RawLibrary>("RPC.GetLibrary", { refresh, force }));
  }

  /**
   * Re-read only what's installed on disk. No round trip to Epic, so it costs a
   * fraction of a full `GetLibrary(true)`: this is the refresh for after an
   * install or an uninstall, where the catalog can't have changed.
   */
  async GetInstalled(): Promise<LibraryResult> {
    return toLibrary(await call<RawLibrary>("RPC.GetLibrary", { installed: true }));
  }

  async GetLaunchCommand(appName: string): Promise<LaunchCommand | undefined> {
    const raw = await call<{ ok: boolean } & Partial<LaunchCommand>>("RPC.GetLaunchCommand", {
      app_name: appName,
    });

    if (!raw.ok || !raw.exe) return undefined;
    return { exe: raw.exe, arguments: raw.arguments ?? "", startDir: raw.startDir ?? "" };
  }

  /**
   * Slow the first time it's asked for a given game - it fetches that game's
   * manifest from Epic - and cached in the backend from then on.
   */
  async GetGameSize(appName: string, refresh = false): Promise<GameSize | undefined> {
    const raw = await call<{ ok: boolean; disk?: number; download?: number; error?: string }>(
      "RPC.GetGameSize",
      { app_name: appName, refresh },
    );

    if (!raw.ok || raw.disk === undefined) {
      logger.debug("Could not size the game", { appName, error: raw.error });
      return undefined;
    }

    return { disk: raw.disk, download: raw.download ?? raw.disk };
  }

  /**
   * Every achievement a game has, with what the account has unlocked. Costs a
   * round trip to Epic the first time it's asked for a given game; `refresh`
   * re-reads one that has gone stale because the game has been played since.
   */
  async GetAchievements(appName: string, refresh = false): Promise<GameAchievements | undefined> {
    const raw = await call<{ ok: boolean; error?: string } & Partial<RawAchievements>>(
      "RPC.GetAchievements",
      { app_name: appName, refresh },
    );

    if (!raw.ok) {
      logger.debug("Could not read achievements", { appName, error: raw.error });
      return undefined;
    }

    return {
      total: raw.total ?? 0,
      unlocked: raw.unlocked ?? 0,
      fetchedAt: raw.fetchedAt ?? 0,
      achievements: asArray(raw.achievements).map(toAchievement),
    };
  }

  /**
   * The achievement counts we already have cached, for every game at once.
   * Never asks Epic, so it's cheap enough to call at startup - which is what
   * the library home's completion sort needs, since it asks about every game.
   */
  async GetCachedAchievements(): Promise<Map<string, AchievementSummary>> {
    const raw = await call<{
      ok: boolean;
      error?: string;
      games?: AchievementSummary[];
    }>("RPC.GetCachedAchievements");

    const summaries = new Map<string, AchievementSummary>();
    if (!raw.ok) {
      logger.debug("Could not read the cached achievements", { error: raw.error });
      return summaries;
    }

    for (const summary of asArray(raw.games)) {
      summaries.set(summary.appName, {
        appName: summary.appName,
        total: summary.total ?? 0,
        unlocked: summary.unlocked ?? 0,
        fetchedAt: summary.fetchedAt ?? 0,
      });
    }

    return summaries;
  }

  /**
   * Returns as soon as the job is spawned - watch it with `GetJobs`. `basePath`
   * is the parent directory, `gameFolder` the directory name inside it,
   * matching legendary's own split.
   */
  async StartInstall(appName: string, basePath?: string, gameFolder?: string) {
    return this.startJob("RPC.StartInstall", {
      app_name: appName,
      base_path: basePath,
      game_folder: gameFolder,
    });
  }

  async StartUninstall(appName: string) {
    return this.startJob("RPC.StartUninstall", { app_name: appName });
  }

  private async startJob(
    route: `RPC.${string}`,
    payload: { app_name: string; base_path?: string; game_folder?: string },
  ) {
    const raw = await call<{ ok: boolean; job?: Job; error?: string }>(route, payload);

    if (!raw.ok || !raw.job) {
      logger.warn(`${route.slice(4)} failed`, { appName: payload.app_name, error: raw.error });
      return undefined;
    }

    return raw.job;
  }

  /** Every install and uninstall the backend knows about. Cheap enough to poll. */
  async GetJobs(): Promise<Job[]> {
    const raw = await call<{ ok: boolean; jobs?: Job[] }>("RPC.GetJobs");
    return asArray(raw.jobs);
  }

  async PauseJob(appName: string) {
    return (await call<{ ok: boolean }>("RPC.PauseJob", { app_name: appName })).ok;
  }

  async CancelJob(appName: string) {
    return (await call<{ ok: boolean }>("RPC.CancelJob", { app_name: appName })).ok;
  }

  /**
   * Rename the icon Steam just wrote into the name it reads icons back from,
   * and return where that ended up. See services/icons.ts.
   */
  async PlaceIcon(appId: number, accountId?: number) {
    const raw = await call<{ ok: boolean; path?: string; error?: string }>("RPC.PlaceIcon", {
      app_id: appId,
      account_id: accountId,
    });

    if (!raw.ok) {
      logger.debug("Could not place the icon", { appId, error: raw.error });
      return undefined;
    }

    return raw.path;
  }
}

export default new RPC();

File: backend/rpc.lua
local grid = require("grid")
local jobs = require("jobs")
local json = require("json")
local legendary = require("legendary")
local library = require("library")
local logger = require("logger")
local text = require("text")
local trace = require("trace")
local utils = require("utils")

-- Every method takes one JSON payload and answers with one JSON document.

--- Polled once a second while a download is going, so tracing every one of
--- these would bury everything else in the file. Only traced when it fails or
--- takes long enough to be worth knowing about.
local QUIET = { GetJobs = true }

--- What counts as long enough, for calls that are otherwise instant.
local SLOW_MS = 250

---Walk a handler's answer and make every string in it encodable. Keys as well
---as values, since a game name can be either.
---@param value any
---@param seen table<table, table>
---@param report { replaced: integer, window: string|nil }
---@return any
local function sanitise(value, seen, report)
  if type(value) == "string" then
    local scrubbed, replaced, position = text.scrub(value)
    if replaced > 0 and position then
      report.replaced = report.replaced + replaced
      report.window = report.window or text.window(value, position)
    end
    return scrubbed
  end
  if type(value) ~= "table" then
    return value
  end
  if seen[value] then
    return seen[value]
  end

  local copy = {}
  seen[value] = copy
  for key, item in pairs(value) do
    copy[sanitise(key, seen, report)] = sanitise(item, seen, report)
  end
  return copy
end

---Note that a call finished.
---@param name string
---@param started integer
---@param fields table<string, any>
local function record(name, started, fields)
  local elapsed = utils.time_ms() - started
  if QUIET[name] and elapsed < SLOW_MS then
    return
  end

  fields.method = name
  fields.ms = elapsed
  trace.event("rpc", fields)
end

---Wrap a handler so it takes and returns JSON.
---
---Nothing here is allowed to throw: an error raised inside a handler unwinds
---into Millennium rather than back to the caller, which loses the call and the
---reason for it. So every half answers with a document instead.
---@param name string
---@param handler fun(data: table): table|nil
---@return fun(payload: string): string
local function method(name, handler)
  return function(payload)
    local started = utils.time_ms()

    local read, data = pcall(json.decode, payload)
    if not read or type(data) ~= "table" then
      trace.anomaly("rpc.payload", { method = name, bytes = #payload })
      return json.encode({ ok = false, error = "Malformed request payload" })
    end

    local ok, result = pcall(handler, data)
    if not ok then
      logger:error("RPC handler failed: " .. tostring(result))
      trace.anomaly("rpc.failed", { method = name, error = tostring(result) })
      return json.encode({ ok = false, error = (text.scrub(tostring(result))) })
    end

    -- Cannot return nothing or nil alone, because Millennium tries to stoi it.
    if not result then
      record(name, started, {})
      return "null"
    end

    local report = { replaced = 0 }
    -- Both halves inside the pcall, not just the encode: an argument is
    -- evaluated before the call that protects it, so sanitising out here would
    -- put its errors exactly where this whole function exists to keep them from
    -- going. Encoding throws on its own as well, for anything cjson can't
    -- represent - a number that isn't finite, a value type with no JSON
    -- equivalent.
    local wrote, encoded = pcall(function()
      return json.encode(sanitise(result, {}, report))
    end)
    if not wrote then
      logger:error("RPC response could not be encoded: " .. tostring(encoded))
      trace.anomaly("rpc.encode", { method = name, error = tostring(encoded) })
      return json.encode({ ok = false, error = "Response could not be encoded" })
    end

    -- Worth a line of its own: this is the shape that used to reach Millennium's
    -- parser and take the whole host down with it.
    if report.replaced > 0 then
      trace.anomaly(
        "rpc.invalid_utf8",
        { method = name, replaced = report.replaced, bytes = report.window }
      )
    end

    record(name, started, { bytes = #encoded })
    return encoded
  end
end

---@class RPC
local RPC = {}

-- Setup -----------------------------------------------------------------------

---Is legendary runnable, and is an Epic account signed in?
RPC.GetStatus = function(data)
  return legendary.get_status(data.refresh == true)
end

---Exchange the code from Epic's redirect page for a stored login. The browser
---half happens in the frontend. We never see the password - legendary trades
---the code for its own tokens and stores them.
RPC.SignIn = function(data)
  local ok, err = legendary.authenticate(data.code)
  if not ok then
    return { ok = false, error = err }
  end

  -- authenticate() already re-read the status to decide whether it worked.
  return { ok = true, status = legendary.get_status() }
end

RPC.SignOut = function()
  legendary.sign_out()
  -- sign_out() already left the cached status current.
  return { ok = true, status = legendary.get_status() }
end

-- Library ---------------------------------------------------------------------

---Everything the account owns, merged with what's installed on disk.
---
---Answers from the cache by default, since a cold `legendary list` takes a few
---seconds on a large account and shouldn't happen behind the user's back. Pass
---`refresh` to read it, `force` to also bypass legendary's own catalog cache,
---or `installed` to re-read just what's on disk.
RPC.GetLibrary = function(data)
  local games, err
  if data.installed then
    games, err = library.refresh_installed()
  elseif data.refresh then
    games, err = library.refresh(data.force == true)
  else
    games = library.get()
  end

  if not games then
    return { ok = false, error = err }
  end
  return { ok = true, games = games, refreshed_at = library.get_refreshed_at() }
end

---The command line Steam should use for one game's shortcut.
---
---It points at legendary, not the game's own executable: legendary waits on the
---game process, so playtime and "Currently playing" work, the shortcut survives
---Epic moving the exe, and it can exist before the game is installed.
RPC.GetLaunchCommand = function(data)
  local binary = legendary.get_binary()
  if not binary then
    return { ok = false }
  end

  return {
    ok = true,
    exe = binary,
    arguments = 'launch "' .. data.app_name .. '"',
    start_dir = binary:match("^(.*)[\\/]") or "",
  }
end

---How much room one game needs on disk, for the "Space Required" Steam shows
---beside the Install button. Kept out of GetLibrary because it costs a manifest
---fetch per game; slow once per game, cached after that.
RPC.GetGameSize = function(data)
  local size, err, still_running = legendary.get_size(data.app_name, data.refresh == true)
  -- Answered rather than waited for: the frontend asks again after retry_in
  -- milliseconds, and everything else keeps working meanwhile.
  if still_running then
    return { pending = true, retry_in = 500 }
  end
  if not size then
    return { ok = false, error = err }
  end
  return { ok = true, disk = size.disk, download = size.download }
end

---One game's achievements, as Epic has them. Costs a round trip to Epic on a
---miss, cached in the backend after that; pass `refresh` to re-read one that
---has gone stale because the game has been played since.
RPC.GetAchievements = function(data)
  local result, err, still_running = legendary.get_achievements(data.app_name, data.refresh == true)
  -- Answered rather than waited for, the same way GetGameSize is.
  if still_running then
    return { pending = true, retry_in = 500 }
  end
  if not result then
    return { ok = false, error = err }
  end

  -- Copied rather than flagged in place: the table is legendary's cache entry,
  -- and an `ok` written into it would be persisted with it.
  return {
    ok = true,
    total = result.total,
    unlocked = result.unlocked,
    fetched_at = result.fetched_at,
    achievements = result.achievements,
  }
end

---What we already know about every game's achievements, counts only, straight
---from the cache. Never touches Epic, so the frontend can seed the whole
---library's completion percentages in one call.
RPC.GetCachedAchievements = function()
  -- A list rather than a map keyed by app name: the frontend camelises every
  -- key it gets back, and an app name is not a key it should be rewriting.
  local games = {}
  for app_name, summary in pairs(legendary.get_cached_achievements()) do
    table.insert(games, {
      app_name = app_name,
      total = summary.total,
      unlocked = summary.unlocked,
      fetched_at = summary.fetched_at,
    })
  end

  return { ok = true, games = games }
end

-- Installs --------------------------------------------------------------------

---Start installing, updating or resuming one game. Returns as soon as the job
---is spawned; progress is read back with GetJobs. Starting a game that is
---already installing returns the running job.
RPC.StartInstall = function(data)
  local job, err = legendary.install(data.app_name, data.base_path, data.game_folder)
  if not job then
    return { ok = false, error = err }
  end
  return { ok = true, job = job }
end

---Update an installed game to the latest build. The same job machinery as an
---install, and the same queue - only one legendary runs at a time.
RPC.StartUpdate = function(data)
  local job, err = legendary.update(data.app_name)
  if not job then
    return { ok = false, error = err }
  end
  return { ok = true, job = job }
end

---Remove a game from disk. The Steam shortcut is left alone, since keeping
---uninstalled games in the library is the point of the plugin.
RPC.StartUninstall = function(data)
  local job, err = legendary.uninstall(data.app_name)
  if not job then
    return { ok = false, error = err }
  end
  return { ok = true, job = job }
end

---Every install or uninstall we know about, running or finished. Polled while
---something is downloading, so it only reads the small files the runner writes.
RPC.GetJobs = function()
  return { ok = true, jobs = jobs.list() }
end

---Stop a running install, keeping what's already downloaded. legendary has no
---pause of its own, so this kills it; StartInstall is the resume.
RPC.PauseJob = function(data)
  return { ok = jobs.pause(data.app_name) }
end

---Stop a job and forget it. What's on disk is left where it is.
RPC.CancelJob = function(data)
  return { ok = jobs.cancel(data.app_name) }
end

-- Artwork ---------------------------------------------------------------------

---Move the icon Steam just wrote into the name Steam reads icons back from.
---The frontend hands Steam the bytes; this is the part Steam leaves undone.
RPC.PlaceIcon = function(data)
  local path, err = grid.place_icon(data.app_id, data.account_id)
  if not path then
    return { ok = false, error = err }
  end
  return { ok = true, path = path }
end

-- Wrapped in one pass rather than at each definition, so a method cannot be
-- added without its JSON handling, and so the name a trace line carries is the
-- one the frontend called.
for name, handler in pairs(RPC) do
  RPC[name] = method(name, handler)
end

return { RPC = RPC }
